# FateID algorithm

FateID is a random forests-based algorithm for the quantification of the fate biases of multipotent progenitors, i. e. the probabilities of differentiating towards particular cell lineages. It utilizes single cell expression data, and was designed to work well with quantitative single-cell RNA-seq data incorporating unique molecular identifiers. FateID requires a gene-by-cell expression matrix as input and groups of cells represeting committed stages (or advanced differentiation stages) of all lineages accessible from the multipotent progenitor population. These so-called target clusters can be defined by a clustering partition, which can also directly be generated by FateID if a list of specific marker genes for all lineages is given.

## Installing

After downloading and unzipping
```
unzip FateID-master.zip 
```

it can be installed from the command line by
```
R CMD INSTALL FateID-master
```

or directly in R from source by
```
install.packages("FateID-master",repos = NULL, type="source")
```
(if R is started from the directory where FateID.zip has been downloaded to; otherwise specify the full path)


Alternatively, install in R directly from github using devtools:
```
install.packages("devtools")
library(devtools)
install_github("dgrun/FateID")
```

## Running a FateID analysis

Load package:
```
library(FateID)
```

See vignette for details and examples:
```
vignette("FateID")
```

## Reference:

Herman JS, Sagar, Gr√ºn D. (2018) FateID infers cell fate bias in
multipotent progenitors from single-cell RNA-seq data. Nat Methods. 2018 May;15(5):379-386. doi: 10.1038/nmeth.4662.